cascada/primitives/rectangle.py

```diff
@@ -0,0 +1,219 @@
+    output_widths = [16, 16, 16, 16]
+    round_keys = None
+
+    @classmethod
+    def set_num_rounds(cls, new_num_rounds):
+        cls.num_rounds = new_num_rounds
+
+    @classmethod
+    def eval(cls, *x):
+        x = list(x)
+
+        for i in range(cls.num_rounds):
+            add_round_key(x, cls.round_keys, i * 4)
+            sub_column(x)
+            shift_row(x)
+            cls.add_round_outputs(x)
+
+        add_round_key(x, cls.round_keys, cls.num_rounds * 4)
+
+        return x
+
+
+class RECTANGLE80KeySchedule(RoundBasedFunction):
+    """RECTANGLE-80 key schedule function."""
+
+    num_rounds = 25
+    input_widths = [16, 16, 16, 16, 16]
+    output_widths = [16 for _ in range(4 * num_rounds + 4)]
+
+    @classmethod
+    def set_num_rounds(cls, new_num_rounds):
+        assert new_num_rounds <= len(RC)
+        cls.num_rounds = new_num_rounds
+        cls.output_widths = [16 for _ in range(4 * new_num_rounds + 4)]
+
+    @classmethod
+    def eval(cls, *master_key):
+        x = list(master_key)
+        round_keys = x[0:4]
+
+        for i in range(cls.num_rounds):
+            t = x[0:4]
+            sub_column(t)
+            x[0] = (x[0] & 0xFFF0) | (t[0] & 0x000F)
+            x[1] = (x[1] & 0xFFF0) | (t[1] & 0x000F)
+            x[2] = (x[2] & 0xFFF0) | (t[2] & 0x000F)
+            x[3] = (x[3] & 0xFFF0) | (t[3] & 0x000F)
+
+            t = x[0]
+            x[0] = RotateLeft(x[0], 8) ^ x[1]
+            x[1] = x[2]
+            x[2] = x[3]
+            x[3] = RotateLeft(x[3], 12) ^ x[4]
+            x[4] = t
+
+            x[0] ^= Constant(RC[i], 16)
+
+            round_keys.append(x[0])
+            round_keys.append(x[1])
+            round_keys.append(x[2])
+            round_keys.append(x[3])
+
+        return round_keys
+
+
+class RECTANGLE128KeySchedule(RoundBasedFunction):
+    """RECTANGLE-128 key schedule function."""
+
+    num_rounds = 25
+    input_widths = [16, 16, 16, 16, 16, 16, 16, 16]
+    output_widths = [16 for _ in range(4 * num_rounds + 4)]
+
+    @classmethod
+    def set_num_rounds(cls, new_num_rounds):
+        assert new_num_rounds <= len(RC)
+        cls.num_rounds = new_num_rounds
+        cls.output_widths = [16 for _ in range(4 * new_num_rounds + 4)]
+
+    @classmethod
+    def eval(cls, *master_key):
+        x = list(master_key)
+        round_keys = []
+
+        for i in range(cls.num_rounds):
+            round_keys.append(x[1])
+            round_keys.append(x[3])
+            round_keys.append(x[5])
+            round_keys.append(x[7])
+
+            t = [x[1], x[3], x[5], x[7]]
+            sub_column(t)
+            x[1] = (x[1] & 0xFF00) | (t[0] & 0x00FF)
+            x[3] = (x[3] & 0xFF00) | (t[1] & 0x00FF)
+            x[5] = (x[5] & 0xFF00) | (t[2] & 0x00FF)
+            x[7] = (x[7] & 0xFF00) | (t[3] & 0x00FF)
+
+            t0 = x[0]
+            t1 = x[1]
+            x[0] = ((t0 << 8) | (t1 >> 8)) ^ x[2]
+            x[1] = ((t1 << 8) | (t0 >> 8)) ^ x[3]
```

Comment:
Where are these OR operations coming from? Can they be replaced with cyclic rotations? As before, cyclic rotations propagate properties better than OR.

Reply:
RECTANGLE 128 key schedule uses 32-bit words instead of 16-bit. Extract allows 32-bit word key schedule implementation so I have rewritten it so it also has rotation now.